bqskit/passes/search/generators/seed.py

```diff
@@ -60,51 +86,57 @@ def gen_initial_layer(
-                `self.seed`.
+            TypeError: If target is not a UnitaryMatrix, StateVector, or
+                StateSystem.
         """
-
         if not isinstance(target, (UnitaryMatrix, StateVector, StateSystem)):
-            raise TypeError(
-                'Expected unitary or state, got %s.' % type(target),
-            )
-
-        if target.dim != self.seed.dim:
-            raise ValueError('Seed dimension mismatch with target.')
+            raise TypeError(f'Expected unitary or state, got {type(target)}.')
 
-        data['seed_seen_before'] = {self.hash_structure(self.seed)}
+        return Circuit(target.num_qudits)
 
-        return self.seed
-
-    def gen_successors(self, circuit: Circuit, data: PassData) -> list[Circuit]:
+    def gen_successors(
+        self,
+        circuit: Circuit,
+        data: PassData,
+    ) -> list[Circuit]:
         """
-        Generate the successors of a circuit node.
+        Generate the successors of a circuit node. If `circuit` is the empty
+        Circuit, seeds of the correct dimension will be used.
 
         Raises:
-            ValueError: If circuit is a single-qudit circuit.
-        """
+            TypeError: If `circuit` is not a Circuit.
 
+            ValueError: If `circuit` is a single-qudit circuit.
+        """
         if not isinstance(circuit, Circuit):
-            raise TypeError('Expected circuit, got %s.' % type(circuit))
-
-        if circuit.num_qudits < 2:
-            raise ValueError('Cannot expand a single-qudit circuit.')
-
-        if circuit.dim != self.seed.dim:
-            raise ValueError('Seed dimension mismatch with circuit.')
+            raise TypeError(f'Expected Circuit, got {type(circuit)}.')
+
+        # If circuit is empty Circuit, successors are seeds
+        if self._is_empty_circuit(circuit):
+            # Check if any seed dim matches, only use those seeds
+            data['seed_seen_before'] = set()
+            useable_seeds = []
+            for seed in self.seeds:
+                if circuit.dim != seed.dim:
+                    continue
+                useable_seeds.append(seed)
+            # If there are no usable seeds, proceed to normal search
+            if len(useable_seeds) > 0:
+                for seed in useable_seeds:
+                    data['seed_seen_before'].add(self.hash_structure(seed))
+                return useable_seeds
 
         # Generate successors
         successors = self.forward_generator.gen_successors(circuit, data)
 
-        removed_count = 0
-        for cycle, op in circuit.operations_with_cycles(reverse=True):
-            removed_count += 1
-            if removed_count > self.num_removed:
-                break
-            copied_circuit = circuit.copy()
-            copied_circuit.pop((cycle, op.location[0]))
-            successors.insert(0, copied_circuit)
+        # Search reverse direction
+        ancestor_circuits = self.remove_atomic_units(circuit)
```

Comment:
Currently it does not. It will only remove single qudit gates that depend on multi-qudit gates that are removed. As synthesis of single-qudit circuits is not supported in the `QSearchSynthesisPass`, I think this might make sense still? I added a bit in the docstring about this.

Reply:
Qsearch supports single qudit circuits with the proper layer generator, check out `bqskit.compiler.compile.build_single_qudit_retarget_workflow`

Reply:
Added a check that will pop the last `self.num_removed` cycles from single qudit circuits.